src/components/post/post.service.ts

```diff
@@ -0,0 +1,313 @@
+import { forwardRef, Inject, Injectable } from '@nestjs/common';
+import { ConfigService } from 'aws-sdk';
+import { node, relation } from 'cypher-query-builder';
+import { DateTime } from 'luxon';
+import {
+  generateId,
+  NotFoundException,
+  ServerException,
+  Session,
+} from '../../common';
+import {
+  createBaseNode,
+  DatabaseService,
+  ILogger,
+  Logger,
+  matchRequestingUser,
+} from '../../core';
+import {
+  calculateTotalAndPaginateList,
+  defaultSorter,
+  matchPropList,
+} from '../../core/database/query';
+import {
+  DbPropsOfDto,
+  parseBaseNodeProperties,
+  parsePropList,
+  runListQuery,
+  StandardReadResult,
+} from '../../core/database/results';
+import { AuthorizationService } from '../authorization/authorization.service';
+import { UserService } from '../user';
+import { CreatePost, Post, UpdatePost } from './dto';
+import { PostListInput, PostListOutput } from './dto/list-posts.dto';
+
+@Injectable()
+export class PostService {
+  private readonly securedProperties = {
+    body: true,
+    user: true,
+    modifiedAt: true,
+  };
+
+  constructor(
+    private readonly db: DatabaseService,
+    private readonly config: ConfigService,
+    @Inject(forwardRef(() => UserService))
+    private readonly userService: UserService,
+    @Inject(forwardRef(() => AuthorizationService))
+    private readonly authorizationService: AuthorizationService,
+    @Logger('post:service') private readonly logger: ILogger
+  ) {}
+
+  async create(
+    { projectId, engagementId, parentPostId, ...input }: CreatePost,
+    session: Session
+  ): Promise<Post> {
+    const createdAt = DateTime.local();
+    const postId = await generateId();
+
+    const secureProps = [
+      {
+        key: 'body',
+        value: input.body,
+        isPublic: false,
+        isOrgPublic: false,
+      },
+      {
+        key: 'modifiedAt',
+        value: createdAt,
+        isPublic: false,
+        isOrgPublic: false,
+      },
+    ];
+
+    if (!projectId && !engagementId && !parentPostId) {
+      throw new ServerException(
+        'A post must be associated with either a project, engagement, or parent post.'
+      );
+    }
+
+    try {
+      const createPost = this.db
+        .query()
+        .call(matchRequestingUser, session)
+        .call(createBaseNode, postId, 'Post', secureProps)
+        .return('node.id as id');
+
+      await createPost.first();
+
+      // connect the Post to the User
+      await this.db
+        .query()
+        .match([
+          [node('user', 'User', { id: session.userId })],
+          [node('post', 'Post', { id: postId })],
+        ])
+        .create([
+          node('post'),
+          relation('out', '', 'user', {
+            active: true,
+            createdAt: DateTime.local(),
+          }),
+          node('user'),
+        ])
+        .return('post.id as id')
+        .first();
```

Comment:
We've gone back and said its better to store the user.id as a prop than to link back to the user because it avoids the user node having many, many relationships coming into it.

Reply:
Done! will push up update shortly.